### PR TITLE
Add hookman generator

### PR DESCRIPTION
**Pull request by williamjamir**

### CODE REVIEW

File: hookman/hook_man_generator.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        with open(hook_caller_hpp, mode='w') as file:
+            file.writelines(dedent(f"""\
+                add_library({self.pyd_name}_interface INTERFACE)
+                target_include_directories({self.pyd_name}_interface INTERFACE ./)
+                """))
+
+        with open(hook_caller_python, mode='w') as file:
+            file.writelines(dedent(f"""\
+            include(pybind11Tools)
+
```
Comment (nicoddemus) [inline hookman/hook_man_generator.py]: Mixed line-endings

File: hookman/hooks.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    - Project Name:
+        This field will be used to identify the project and to name the hook functions
+    - Version:
+        The current version of the spec, when a new hook is created or modified this version should be incremented
+    - Hooks:
+        A list with the hooks available for the project, each hook is represented by a python function
+    """
+
+    def __init__(self, *, project_name: str, version: str, pyd_name: str, hooks: List[Callable]) -> None:
+        [self._check_hook_arguments(hook) for hook in hooks]
```
Comment (nicoddemus) [inline hookman/hooks.py]: It is considered bad style to use a list comprehension just for the implicit loop, and discard the results. Just loop over it instead:

```python
for hook in hooks:
    self._check_hook_arguments(hook) 
```

File: hookman/hooks.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        if not hook_args.args:
+            raise TypeError("It's not possible to create a hook without argument")
+
+        annotate_args = {
+            arg: hook_args.annotations[arg]
+            for arg in hook_args.annotations
+            if arg != 'return'
+        }
+
+        if len(annotate_args) != len(hook_args.args):
```
Comment (nicoddemus) [inline hookman/hooks.py]: 👍

File: hookman/hooks.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        self.hooks_available = {
+            f'{hook.__name__.lower()}': f'{specs.project_name}_v{specs.version}_{hook.__name__.lower()}'
+            for hook in specs.hooks
+        }
+
+    def get_hook_caller(self) -> 'HookCaller':
+        """
+        Return a HookCaller class that holds all references for the functions implemented on the plugins.
+        """
+        # self._spec.pyd_name = '_alfasim_hookcaller'
```
Comment (nicoddemus) [inline hookman/hooks.py]: commented code

File: hookman/hooks.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        return dll_locations
+
+    def _bind_dll_functions_on_hook_caller(self, dll_path: Path, hook_caller : 'HookCaller'):
+        """
+        Load the dll_path from the plugin and bind methods that are implemented on the hook_caller
+        """
+        plugin_dll = ctypes.cdll.LoadLibrary(str(dll_path))
+
+        hooks_to_bind = {
```
Comment (nicoddemus) [inline hookman/hooks.py]: So this is binding the *last* plugin which implements a certain function, is that what you intended? I remember we deciding it should raise an error if more than one plugin implemented a hook function.

File: setup.py [hunk 6]
```diff
@@ -21,15 +21,9 @@
     author="ESSS",
     author_email='foss@esss.co',
     classifiers=[
-        'Development Status :: 2 - Pre-Alpha',
```
Comment (nicoddemus) [inline setup.py]: I think we should advertise it as `'Development Status :: 3 - Alpha'`

File: tasks.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+            """))
+
+        list_with_c_files_names = [c_file for c_file in hook_spec_path.parent.glob('plugin/*.c')]
+        for i in list_with_c_files_names:
+            shutil.copy2(src=i, dst=dir_for_compilation / 'plugin')
+        cmake_plugin = dir_for_compilation / 'plugin/CMakeLists.txt'
+        if list_with_c_files_names:
+            with open(cmake_plugin, mode='w') as file:
+                file.writelines(dedent(f"""\
+                    add_library({folder_test_name} SHARED {" ".join(str(x.name) for x in list_with_c_files_names)} hook_specs.h)
```
Comment (nicoddemus) [inline tasks.py]: Mixed end of lines

File: tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+import pytest
+
+
+
+@pytest.fixture
+def libs_path():
+    return Path(__file__).parents[1] / 'build/libs'
+
+
+# @pytest.fixture
```
Comment (nicoddemus) [inline tests/conftest.py]: Commented code

File: tests/test_hook_man_generator/ExpectedHookCaller.hpp [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace HookMan {
+
+template <typename F_TYPE> std::function<F_TYPE> register_c_func(uintptr_t p) {
+    return std::function<F_TYPE>(reinterpret_cast<F_TYPE *>(p));
+}
+
+class HookCaller {
+public:
+    inline int friction_factor(int v1, int v2) { return this->_friction_factor(v1, v2); }
```
Comment (nicoddemus) [inline tests/test_hook_man_generator/ExpectedHookCaller.hpp]: How should we handle default functions for hooks?

File: tests/test_hooks.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+    HookMan(specs=plugin_specs, plugin_dirs=[datadir])
+
+    with pytest.raises(FileNotFoundError, match="The given path doesn't have a .yaml file"):
+        HookMan(specs=plugin_specs, plugin_dirs=[Path('NON_EXISTING_PATH')])
+
+
+def test_check_hook_arguments():
+
+    def test_method_without_arguments() -> 'float':
```
Comment (nicoddemus) [inline tests/test_hooks.py]: I suggest to write 3 separate tests, instead of bundling them together here.

File: tests/test_hooks.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        specs = HooksSpecs(project_name='acme', version='1', pyd_name='_acme', hooks=[fail_method_with_docs_missing])
+
+
+def test_get_hook_caller(datadir, libs_path, plugin_specs):
+    simple_plugin_dll = libs_path / 'test_hooks.dll'
+    dst_path = datadir / 'plugin'
+    copy2(src=simple_plugin_dll, dst=dst_path)
+
+    hm = HookMan(specs=plugin_specs, plugin_dirs=[datadir])
+    hookCaller = hm.get_hook_caller()
```
Comment (nicoddemus) [inline tests/test_hooks.py]: I think you should actually test that the hook caller is correct bound. Also, mind the naming of `hookCaller`, should be `hook_caller`.

File: tests/test_hooks.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    # A arguments of the hook must inform the type
+    with pytest.raises(TypeError, match="All hooks arguments must have the type informed"):
+        specs = HooksSpecs(project_name='acme', version='1', pyd_name='_acme', hooks=[fail_method_with_missing_type_on_argument])
+
+    with pytest.raises(TypeError, match="All hooks must have documentation"):
+        specs = HooksSpecs(project_name='acme', version='1', pyd_name='_acme', hooks=[fail_method_with_docs_missing])
+
+
+def test_get_hook_caller(datadir, libs_path, plugin_specs):
+    simple_plugin_dll = libs_path / 'test_hooks.dll'
```
Comment (nicoddemus) [inline tests/test_hooks.py]: `.dll` even on Linux?

File: tests/test_hooks/plugin/simple_plugin.c [hunk 13]
```diff
@@ -0,0 +1,7 @@
+#include "hook_specs.h"
+
+HOOK_FRICTION_FACTOR(value1, value2) {
+        return value1 + value2;
+    }
```
Comment (nicoddemus) [inline tests/test_hooks/plugin/simple_plugin.c]: Double indentation here

File: tox.ini [hunk 14]
```diff
@@ -4,7 +4,7 @@ envlist = py36, flake8
 [travis]
 python =
```
Comment (nicoddemus) [inline tox.ini]: This file is not being used right?

File: hookman/cli.py [hunk 15]
```diff
@@ -6,11 +6,33 @@
...
+    Per default dst-path is the same directory that the command is called.
+
+    Example:
+    > hookman /<some_dir>/hook_specs.py --dst-path=/home/<some_other_path>
+
+    """
+
+    from pathlib import Path
+    hook_specs_path = Path(specs_path)
+    if Path(hook_specs_path).is_file():
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/cli.py]: - [ ] Remove the file check

File: hookman/cli.py [hunk 16]
```diff
@@ -6,11 +6,33 @@
...
+        from hookman.hook_man_generator import HookManGenerator
+        hm_generator = HookManGenerator(hook_spec_file_path=hook_specs_path)
+        hm_generator.generate_files(Path(dst_path))
+    else:
+        return
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/cli.py]: The check was removed, now the `HookManGenerator` raises `RuntimeError`

[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/cli.py]: The `HookManGenerator` raises a RuntimeError("Invalid file, specs not defined.")

[the remark this answers is not in the document]
Comment (nicoddemus) [reply hookman/cli.py]: see `if Path(hook_specs_path).is_file():` above ;)

File: hookman/hook_man_generator.py [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+    """
+    Class to assist on the process to generate the files
+
+    name: Name of the Hook
+    macro_name: Name of the Hook in Upper Case
+    r_type: Type of the return from the hook
+
+    args: The name of each argument
+        Ex.: v1, v2, v3
+    args_type: The type of each argument
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: Currently, I'm using NamedTuple just to keep this list of parameter organized If you find it necessary I can change the NamedTuples to attrs (or dataclass).
What do you think?

File: hookman/hook_man_generator.py [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+        Ex.: int, float, int
+    args_with_type: The name of the argument with the type
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: - [x] Checked

File: hookman/hook_man_generator.py [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+        Ex.: int v1, float v2, int v3
+
+    function_name: Full name of the hook function
+        Ex.: project_name_version_hook_name -> alfasim_v4_friction_factory
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: - [x] Remove the file check

File: hookman/hook_man_generator.py [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+        """
+        Receives a path to a hooks specification file.
+        if the Path provided is not a file an exception FileNotFoundError is raised.
+        If the File provided doesn't have a spec object, a RuntimeError is raised.
+        """
+        if hook_spec_file_path.is_file():
+            hook_spec_module = self._import_hook_specification_file(hook_spec_file_path)
+            self._populate_local_variables(hook_spec_module.specs)
+        else:
+            raise FileNotFoundError("You must inform a file.")
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: - [X] Applied

File: hookman/hook_man_generator.py [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+
+    def _import_hook_specification_file(self, hook_spec_file_path: Path) -> HooksSpecs:
+        """
+        Returns the "HooksSpecs" object that defines the hook specification provide from the project.
+        The file is considered valid if the importlib can access the object called "specs"
+
+        :param hook_spec_file_path: Path to the location of the file
+        :return: A Python module that represent specification from the given file
+        """
+        spec = importlib.util.spec_from_file_location('hook_specs', hook_spec_file_path)
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: The first argument of `spec_from_file_location` is the name that the module will receive (in this case the module will be named `"hook_specs"`

Currently, I'm not using this information (the name), I'm just getting the object "specs" inside this module.

File: hookman/hook_man_generator.py [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+        for hook_spec in hook_specs.hooks:
+            hook_arg_spec = inspect.getfullargspec(hook_spec)
+            hook_arguments = hook_arg_spec.args
+            hook_types = hook_arg_spec.annotations
+            self.hooks.append(
+                Hook(
+                    name=hook_spec.__name__.lower(),
+                    macro_name=hook_spec.__name__.upper(),
+                    r_type=hook_arg_spec.annotations['return'],
+                    args=format_arguments(hook_arguments),
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: - [ ] Check what needs to be done

File: hookman/hook_man_generator.py [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+        template <typename F_TYPE> std::function<F_TYPE> register_c_func(uintptr_t p) {{
+        {INDENTATION}return std::function<F_TYPE>(reinterpret_cast<F_TYPE *>(p));
+        }}
+
+        class HookCaller {{
+        public:
+        """)
+
+        list_with_hook_calls += [
+            f'{INDENTATION}inline {hook.r_type} {hook.name}({hook.args_with_type}) {{ return this->_{hook.name}({hook.args}); }}' + NEW_LINE
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: - [x] Created https://github.com/ESSS/hookman/issues/2

File: hookman/hooks.py (hunk 2 above)
Comment (williamjamir) [reply hookman/hooks.py]: - [x] Applied

File: hookman/hooks.py (hunk 4 above)
Comment (williamjamir) [reply hookman/hooks.py]: - [X] Remove commented code

File: setup.py (hunk 6 above)
Comment (williamjamir) [reply setup.py]: - [x] Add 'Development Status :: 3 - Alpha`

File: tasks.py (hunk 7 above)
Comment (williamjamir) [reply tasks.py]: - [ ] Normalize all the line endings

File: tests/conftest.py (hunk 8 above)
Comment (williamjamir) [reply tests/conftest.py]: - [x] Remove

File: tests/test_hook_man_generator/ExpectedHookCaller.hpp (hunk 9 above)
Comment (nicoddemus) [reply tests/test_hook_man_generator/ExpectedHookCaller.hpp]: As we discussed:

```c++
std::function<int(int, int)> friction_factor() {
    reutnr this->_friction_factor;
}
```

File: tox.ini (hunk 14 above)
Comment (williamjamir) [reply tox.ini]: - [x] Fixed

File: tests/test_hooks/plugin/simple_plugin.c (hunk 13 above)
Comment (williamjamir) [reply tests/test_hooks/plugin/simple_plugin.c]: - [x] Fixed

File: tests/test_hooks.py (hunk 12 above)
Comment (williamjamir) [reply tests/test_hooks.py]: Commit https://github.com/ESSS/hookman/pull/1/commits/be40b6952c8bba84881bb0d6ac544518916d8a8e fix this issue

File: hookman/hook_man_generator.py (hunk 1 above)
Comment (williamjamir) [reply hookman/hook_man_generator.py]: Commit https://github.com/ESSS/hookman/pull/1/commits/b2980dd1efe4669c3bb69b739a065c3f5f17997e normalize the line endings

File: hookman/hooks.py (hunk 5 above)
Comment (williamjamir) [reply hookman/hooks.py]: This logic of "if more than one plugin implemented a hook function" I was planning to do with the next task (to list all implemented hooks and so on).

File: tests/test_hook_man_generator/ExpectedHookCaller.hpp (hunk 9 above)
Comment (williamjamir) [reply tests/test_hook_man_generator/ExpectedHookCaller.hpp]: - [x] Applied

File: tests/test_hooks.py (hunk 10 above)
Comment (williamjamir) [reply tests/test_hooks.py]: - [x] separate in 3 tests

File: tests/test_hooks.py (hunk 11 above)
Comment (williamjamir) [reply tests/test_hooks.py]: - [x] Fixed on the last commit

File: .travis.yml [hunk 24]
```diff
@@ -1,15 +1,33 @@
...
 python:
   - 3.6
 
 
 # Command to install dependencies, e.g. pip install -r requirements.txt --use-mirrors
-install: pip install -U pip tox-travis codecov codacy-coverage
+install:
+  - pip install -U pip tox-travis codecov codacy-coverage
+
+  - wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O miniconda.sh
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply .travis.yml]: I'm just using conda now on travis (and appveyor) to have access to the CMake configurations for `Pybind11`

File: hookman/hook_man_generator.py [hunk 25]
```diff
@@ -0,0 +1,265 @@
...
+        for hook in self.hooks:
+            line = f'#define HOOK_{hook.macro_name}({hook.args}) API_EXP {hook.r_type} __cdecl {hook.function_name}({hook.args_with_type})' + NEW_LINE
+            list_with_hook_specs_arguments.append(line)
+
+        from textwrap import dedent
+        file_content += dedent(f"""\
+        #ifndef HEADER_FILE
+        #define HEADER_FILE
+        #ifdef WIN32
+            #define API_EXP	__declspec(dllexport)
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: - [x] fixed

File: hookman/hook_man_generator.py [hunk 26]
```diff
@@ -0,0 +1,265 @@
...
+        """
+        file_content = []
+        list_with_hook_specs_arguments = []
+
+        for hook in self.hooks:
+            line = f'#define HOOK_{hook.macro_name}({hook.args}) API_EXP {hook.r_type} __cdecl {hook.function_name}({hook.args_with_type})' + NEW_LINE
+            list_with_hook_specs_arguments.append(line)
+
+        from textwrap import dedent
+        file_content += dedent(f"""\
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply hookman/hook_man_generator.py]: Commit b2980dd normalize the line endings

File: tests/test_hook_man_generator.py [hunk 27]
```diff
@@ -0,0 +1,34 @@
+from pathlib import Path
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply tests/test_hook_man_generator.py]: Fixed on PR https://github.com/ESSS/hookman/pull/3

File: tests/test_hook_man_generator/ExpectedHookCaller.hpp [hunk 28]
```diff
@@ -0,0 +1,22 @@
+#include <functional>
+
+namespace HookMan {
+
+template <typename F_TYPE> std::function<F_TYPE> register_c_func(uintptr_t p) {
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply tests/test_hook_man_generator/ExpectedHookCaller.hpp]: Fixed on PR https://github.com/ESSS/hookman/pull/3

File: tests/test_hook_man_generator/ExpectedHookCaller.hpp [hunk 29]
```diff
@@ -0,0 +1,22 @@
+#include <functional>
+
+namespace HookMan {
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply tests/test_hook_man_generator/ExpectedHookCaller.hpp]: Fixed on PR https://github.com/ESSS/hookman/pull/3

File: tests/test_hook_man_generator/expected_hook_specs.h [hunk 30]
```diff
@@ -0,0 +1,14 @@
+#ifndef HEADER_FILE
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply tests/test_hook_man_generator/expected_hook_specs.h]: Fixed on PR https://github.com/ESSS/hookman/pull/3
I'm using <PROJECT_NAME> at the beginning

File: tests/test_hooks/plugin/simple_plugin.c [hunk 31]
```diff
@@ -0,0 +1,7 @@
+#include "hook_specs.h"
+
+HOOK_FRICTION_FACTOR(value1, value2) {
+        return value1 + value2;
```
[the remark this answers is not in the document]
Comment (williamjamir) [reply tests/test_hooks/plugin/simple_plugin.c]: Fixed on PR https://github.com/ESSS/hookman/pull/3

**Pull request merged**